tests/proptest_tests.rs

```diff
@@ -717,6 +717,10 @@ proptest! {
     fn pt_lines_iter_exact_01(idx in 1024usize..(CHAR_LEN - 1024)) {
         let r = Rope::from_str(TEXT);
         let s = r.slice(idx..(idx + 373));
+        // newline terminator must never be split
+        if s.char(372) == '\r'{
+            return Ok(())
+        }
```

Comment:
This code addition looks incorrect to me.  There's nothing wrong with slicing a rope in the middle of a CRLF pair.  Moreover, stand-alone CR's are perfectly okay and can occur.  So this is a case that actually should be tested, not skipped.

Reply:
Sorry I meant to add a comment and forgot. I am not sure if I ran into a bug in ropey here or how to handle this.
Without this line, this tests failed because the `size_hint` of the lines iterator under reported the line by one (so the iterator did one extra iteration and caused an iteger underflow).
However the size_hint simply reports the number of lines returned by `slice.len_lines()` (and that is check with an assert in this test). So the problem is that slicing in the middle of a CRLF pair causes `RopeSlice::len_lines()` to return the wrong value.
I assumed that you were not supposed to slice in the middle of a CRLF pair (since chunks are explicitly not split in the middle of CRLF pairs). If that is not a case I guess its an (unrelated) bug

Reply:
> So the problem is that slicing in the middle of a CRLF pair causes RopeSlice::len_lines() to return the wrong value.

Ah!  Thanks for the bug catch.  Indeed, slicing fails to account for splitting a CRLF pair at its end.  Working on a fix now.

Reply:
Fixed in 8699de0908b3853431e2dcac6eb70faad7f326d0

Reply:
Also, the additional unit tests I added in the fix revealed that the current lines iterator doesn't handle this case properly.  So your PR here will not only make `Lines` way faster, but also more correct! :-)

Reply:
your tests exposed a couple extra edgecases so its really great we cought this :) I have pushed fixes (see comment below)